### PR TITLE
[OPS-2310] Sidekiq parallel

### PR DESCRIPTION
**Pull request by louman**

### CODE REVIEW

File: lib/patches/tenant_worker.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+require 'sidekiq'
+
+class Patches::TenantWorker
+  include Sidekiq::Worker
+  include Patches::TenantRunConcern
+
+  sidekiq_options Patches::Config.configuration.sidekiq_options
```
Comment (nsinenko) [inline lib/patches/tenant_worker.rb]: Can this just be added to `Patches::TenantRunner`?

File: lib/patches/tenant_runner.rb [hunk 2]
```diff
@@ -8,19 +9,22 @@ def initialize(path: nil, tenants: nil)
...
     Patches.logger.info("Patches tenant runner for: #{tenants.join(',')}")
-
     tenants.each do |tenant|
-      Apartment::Tenant.switch(tenant)
-      runner = build
-      runner.perform
+      if parallel?
+        Patches::TenantWorker.perform_async(tenant, path)
+      else
+        run(tenant, path)
```
Comment (nsinenko) [inline lib/patches/tenant_runner.rb]: Just my 2c, I'd rather avoid using Concerns (a Ruby module in this case) just for the sake of making classes smaller. This is purely a cosmetic change and adds a cognitive overhead finding methods scattered across multiple files.

File: spec/tenant_worker_spec.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+require 'spec_helper'
+require "patches/tenant_worker"
+
+describe Patches::TenantWorker do
+  describe '#perform' do
+    specify do
+      is_expected.to receive(:run).with('test', 'path')
+      subject.perform('test', 'path')
+      # expect(Apartment::Tenant).to receive(:switch).with('test')
+      # expect(Patches::Runner).to receive(:new).with('path')
```
Comment (nsinenko) [inline spec/tenant_worker_spec.rb]: Is this meant to be commented out?

File: lib/patches/tenant_runner.rb (hunk 2 above)
Comment (louman) [reply lib/patches/tenant_runner.rb]: It's not being used for the sake of making the class smaller but re-using the method.

File: lib/patches/tenant_worker.rb (hunk 1 above)
Comment (louman) [reply lib/patches/tenant_worker.rb]: We need to have a new job enqueued for each tenant.

File: spec/tenant_worker_spec.rb (hunk 3 above)
Comment (louman) [reply spec/tenant_worker_spec.rb]: whoops, removed.

File: lib/patches/tenant_runner.rb (hunk 2 above)
Comment (nsinenko) [reply lib/patches/tenant_runner.rb]: Hmm why does it need to be reused? Wouldn't this work?
```ruby
if parallel?
  Patches::TenantWorker.perform_async(tenant, path)
else
  Patches::TenantWorker.new.perform(tenant, path)
end
```

Comment (louman) [reply lib/patches/tenant_runner.rb]: We can't use the worker for both cases cause Sidekiq is not a mandatory dependency. You should be able to use patches without it. At least in theory given the check for defined?(Sidekiq) that I found at some point.

File: spec/tenant_runner_spec.rb [hunk 4]
```diff
@@ -21,16 +24,37 @@ module Tenant
...
+    before { expect(Apartment).to receive(:tenant_names).and_return(tenant_names) }
 
     specify do
       expect(subject.tenants).to eql(['test'])
-      expect(runner).to receive(:perform).and_return(true)
-      expect(subject).to receive(:build).and_return(runner)
-      expect(Apartment::Tenant).to receive(:switch).with('test')
-      expect(subject.perform).to eql(['test'])
+      # expect(Patches::TenantWorker).to receive(:perform).with('test', nil).and_call_original
+      # expect(Apartment::Tenant).to receive(:switch).with('test')
```
Comment (nsinenko) [inline spec/tenant_runner_spec.rb]: Comments :)

**Comment by nsinenko** [comment]
👍

File: spec/config_spec.rb [hunk 5]
```diff
@@ -55,6 +55,15 @@
     end
   end
 
+  describe '#sidekiq_parallel' do
+    subject { patches_config.sidekiq_parallel }
+    before { patches_config.sidekiq_parallel = true }
+
+    it 'is the configured value' do
+      expect(subject).to eq(true)
+    end
```
Comment (s01ipsist) [inline spec/config_spec.rb]: should we also be validating the result if config is not set?

File: lib/patches/tenant_worker.rb (hunk 1 above)
Comment (andrba) [reply lib/patches/tenant_worker.rb]: just learnt that we can configure `sidekiq_options`. Couldn't find it anywhere in the docs, is it worth mentioning it?

File: docs/usage.md [hunk 6]
```diff
@@ -38,6 +38,19 @@ Patches::Config.configure do |config|
 end
 ```
 
+### Running specs in parallel for tenants
```
[the remark this answers is not in the document]
Comment (louman) [reply docs/usage.md]: whoops

File: lib/patches/tenant_worker.rb (hunk 1 above)
Comment (louman) [reply lib/patches/tenant_worker.rb]: good idea, added to the note in the usage instructions.

File: spec/config_spec.rb (hunk 5 above)
Comment (louman) [reply spec/config_spec.rb]: done

**Pull request merged**